File: matlab/CovarianceMatrix.m
classdef CovarianceMatrix < handle
    % CovarianceMatrix   Representation of a covariance matric C for use in 
    %                    generalized least squares problems.
    %                    We denote W'*W = inv(C)
    %
    % CovarianceMatrix methods:
    %    weigh    - Multiplies a given matrix or column vector by W
    %    explicit - Returns an explicit dense covariance matrix C 
    %
    % Constructors:
    %    CovarianceMatrix(Z,type) where Z is a representation of C and type
    %                             defines how Z is related to C.
    %
    % type=='C' (covariance):           Z=C,         L=chol(C,'lower'), W=inv(L)
    % type=='I' (inverse covariance):   Z=inv(C),    L=chol(C,'lower')  W=L'
    % type=='M' (sqrt of covariance):   Z*Z'=C, e.g. Z=chol(C,'lower')  W=inv(Z)
    % type=='W' (inverse factor):       Z'*Z=inv(C),                    W=Z
    % type=='w' (diagonal inv. factor): diag(Z)*diag(Z)=inv(C),                    W=Z
    %
    % Copyright 2020-2024 Sivan Toledo, Tel Aviv University.

 
    properties
        type; 
        invW;
        W;
        w;
        C; % for lazy factorization
    end

    methods 
        function cov = CovarianceMatrix(Z,type)
            %CovarianceMatrix constructor
            %
            % cov = CovarianceMatrix(Z,type)
            %    Z is a real matrix or vector
            %    type describes how Z represents a covariance matrix C
            %    'C': C = Z
            %    'I': C = inv(Z)
            %    'W': C = inv( Z' * Z )
            %    'w': C = diag( Z.^-2 ) = inv( diag(Z)' * diag(Z) )
            %    'M': C = Z' * Z
            %    'U': C = Z * Z' and Z is upper triangular
            %
            cov = cov@handle();
            switch type
                case 'C'       % an explicit covariance matrix
                    % Sivan July 2024, lazy factorization for conventional
                    % filters.
                    %cov.type = 'U';
                    %cov.invW = chol(Z);
                    cov.type = 'C';
                    cov.C    = Z;
                case 'I'       % the inverse of the covariance matrix 
                    cov.type = 'W';
                    cov.W = (chol(Z,'lower'))';
                case 'M'       % a factor (perhaps Cholesky factor) of the covariance matrix 
                    cov.type = 'M'; % hopefully never happens...
                    if (istriu(Z)); cov.type = 'U'; end;
                    cov.invW = Z;
                case 'W'       % an inverse factor W such that W'*W=C
                    cov.type = 'W';
                    cov.W = Z;
                case 'w'       % an inverse factor W such that W'*W=C
                    cov.type = 'w';
                    cov.w = Z;
                otherwise
                    type
                    error('illegal covariance-matrix representation');
            end
        end

        function WA = weigh(cov,A)
            switch cov.type
                case {'U', 'M', 'F'}       
                    WA = (cov.invW)\A;
                case 'W'
                    WA = (cov.W)*A;
                case 'w'
                    WA = diag(cov.w)*A;
                case 'C'
                    cov.type = 'F'; % lowe-Cholesky factored
                    L = chol(cov.C,'lower'); 
                    cov.invW = L; 
                    WA = (cov.invW)\A;
                otherwise
                    error('illegal covariance type');

            end
        end

        function [R,type] = rep(cov)
            type = cov.type;
            switch cov.type
                case {'U', 'M'}
                    R = cov.invW;
                case 'W'
                    R = cov.W;
                case 'w'
                    R = cov.w;
                case 'C'
                    R = cov.C;
                case 'F'
                    % return just one, for the mex interface
                    %R = { cov.C, cov.invW };    
                    R = cov.invW;    
                otherwise
                    error('illegal covariance type');
            end
        end

        function C = explicit(cov)
            switch cov.type
                case {'U', 'M'}
                    C = (cov.invW)' * (cov.invW);
                case 'W'
                    C = inv( (cov.W)' * (cov.W) );
                case 'w'
                    C = diag(cov.w.^-2);
                case {'C', 'F'}
                    C = cov.C;
                otherwise
                    error('illegal covariance type');
            end
        end

    end
end

File: python/covariance_matrix.py
import numpy as np
from scipy.linalg import cholesky, inv

class CovarianceMatrix:
    def __init__(self, Z, matrix_type):
        """
        Initialize a CovarianceMatrix object.

        :param Z: Matrix representing the covariance in various forms.
        :param matrix_type: String specifying the type of representation ('C', 'I', 'W', 'w', 'M', 'U').
        """
        self.type = matrix_type
        self.invW = None
        self.W = None
        self.w = None
        self.C = None  # For lazy factorization

        if matrix_type == 'C':
            self.type = 'C'
            self.C = Z
        elif matrix_type == 'I':
            self.type = 'W'
            self.W = cholesky(Z, lower=True).T
        elif matrix_type == 'M':
            self.type = 'M' if not np.allclose(Z, np.triu(Z)) else 'U'
            self.invW = Z
        elif matrix_type == 'W':
            self.type = 'W'
            self.W = Z
        elif matrix_type == 'w':
            self.type = 'w'
            self.w = Z
        else:
            raise ValueError("Illegal covariance-matrix representation")

    def weigh(self, A):
        """
        Multiply a given matrix or column vector by W.

        :param A: Input matrix or vector.
        :return: Weighted matrix or vector.
        """
        if self.type in {'U', 'M', 'F'}:
            return np.linalg.solve(self.invW, A)
        elif self.type == 'W':
            return self.W @ A
        elif self.type == 'w':
            #print( type(self.w))
            #print( self.w )
            #print (self.w.ravel() )
            #print( self.w.shape )
            #print( self.w.ravel().shape )
         
            #print( type(A))
            #print( A )
            #print( A.shape )
            # ravel() forces self.w to be a 1D array (dimensions (1,1), not (1,) when it is a scalar)
            return np.diag(self.w.ravel()) @ A
        elif self.type == 'C':
            self.type = 'F'  # Lower Cholesky factorized
            self.invW = cholesky(self.C, lower=True)
            return np.linalg.solve(self.invW, A)
        else:
            raise ValueError("Illegal covariance type")

    def rep(self):
        """
        Return the matrix representation of the covariance.

        :return: Tuple (Matrix, Type)
        """
        if self.type in {'U', 'M'}:
            return self.invW, self.type
        elif self.type == 'W':
            return self.W, self.type
        elif self.type == 'w':
            return self.w, self.type
        elif self.type == 'C':
            return self.C, self.type
        elif self.type == 'F':
            return self.invW, self.type
        else:
            raise ValueError("Illegal covariance type")

    def explicit(self):
        """
        Return the explicit dense covariance matrix C.

        :return: Explicit covariance matrix.
        """
        if self.type in {'U', 'M'}:
            return self.invW.T @ self.invW
        elif self.type == 'W':
            return inv(self.W.T @ self.W)
        elif self.type == 'w':
            #print( self.w.flatten().shape )
            #print( self.w )
            #print( self.w.flatten() ** -2 )
            #print( np.diag( self.w.flatten() ** -2 ))
            return np.diag(self.w.flatten() ** -2)
        elif self.type in {'C', 'F'}:
            return self.C
        else:
            raise ValueError("Illegal covariance type")
